keps/sig-cli/3805-ssa-default/README.md

```diff
@@ -0,0 +1,567 @@
+
+Alpha is the current level of the feature since server-side apply is
+currently enabled by default in Kubernetes, but enabled on-demand by
+kubectl users.
+
+The feature already has a fair amount of usage since tools (sometimes
+outside of kubectl) have used it both as "clients" and in controllers.
+
+#### Beta
+
+<!--
+To be re-evaluated later:
+Server-Side Apply has a very limited set of bugs or feature requests as
+this point and is definitely mature. Enabling client-side will allow
+increased usage and reduce burden cost for kubectl to maintain both
+mechanisms. -->
+
+#### GA
+
+<!--
+To be re-evaluated later:
+Kubectl doesn't have real-time metrics for usage. The decision to move
+to server-side entirely by default (if ever enabled) will be driven by
+bug reports and complaints from customers. Also by the ability to
+migrate existing client-side usage to server-side. -->
+
+#### Deprecation
+
+We are not intending to deprecate the flag, but we might remove the
+`--server-side` flag in the long term.
+
+Same thing applies for `--save-config` and other client-side related
+flags in kubectl which we might remove.
+
+### Upgrade / Downgrade Strategy
+
+While upgrade / downgrade doesn't really apply to a kubectl feature, we
+currently have a upgrade (and somewhat downgrade) feature in kubectl to
+go from client-side to server-side apply. The upgrade and downgrade
+works well in the nominal cases but fail with special cases. Enabling
+server-side by default also intends to address that problem.
+
+### Version Skew Strategy
+
+N/A.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Other
+  - Describe the mechanism:
+
+Feature is enabled on-demand, and can stop using it at any time. It also
+requires the `KUBECTL_AUTO_SERVER_SIDE` environment variable to be set.
+
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+
+Feature is client-side only, no.
+
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+Feature is client-side only, no.
+
+###### Does enabling the feature change any default behavior?
+
+No, enabling the `auto` feature means actively selecting it.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+The feature can be disabled by unsetting the `KUBECTL_AUTO_SERVER_SIDE` environment variable.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Toggling the flag value can have impact, but that's already the case.
+Changing the value to an automatic detection will actually help with
+that.
+
+###### Are there any tests for feature enablement/disablement?
+
+Since enablement is not done through a feature gate, tests are fairly
+easy to implement.
+
+### Rollout, Upgrade and Rollback Planning
```

Comment:
nit: it seems like this section is filled out with the default switch in mind, and the previous section seems to be filled out about auto mode. It might be worth saying so at the beginning of each if I'm right about that.

Reply:
reworded